delivery/client.py

```diff
@@ -109,9 +123,93 @@ class DeliveryServiceClient:
     def __init__(
         self,
         routes: DeliveryServiceRoutes,
+        cfg_factory=None,
     ):
         self._routes = routes
+        self.cfg_factory = cfg_factory or ci.util.ctx().cfg_factory()
         self.session = requests.sessions.Session()
+        self.bearer_token = None
+
+    def _authenticate(self):
+        if self.bearer_token:
+            return
+
+        res = self.session.get(
+            url=self._routes.auth_configs(),
+            timeout=(4, 31),
+        )
+
+        res.raise_for_status()
+
+        auth_configs = res.json()
+
+        api_urls = [
+            auth_config.get('api_url')
+            for auth_config in auth_configs
+        ]
+
+        for github_cred in self.cfg_factory._cfg_elements('github'):
+            if github_cred.api_url() not in api_urls:
+                continue
+            break
+        else:
+            logger.info(
+                f'did not find valid github credentials for one of {api_urls=} - '
+                'attempting anonymous-auth'
+            )
+            return
+
+        params = {
+            'access_token': github_cred.credentials().auth_token(),
+            'api_url': github_cred.api_url(),
+        }
+
+        res = self.session.get(
+            url=self._routes.auth(),
+            params=params,
+            timeout=(4, 31),
+        )
+
+        if not res.ok:
+            logger.warning(
+                'authentication against delivery-service failed: '
+                f'{res.status_code=} {res.reason=} {res.content=}'
+            )
+
+        res.raise_for_status()
+
+        self.bearer_token = res.cookies.get(jwt_key)
+
+    def _request(
```

Comment:
nit: considering with this change, `_session` should no longer be used directly, I think it might be more intuitive to make this method "public" (remove leading underscore) and the session-object "private", or make both public (+ add docstr explaining when self.session vs self.request ought to be used).